### PR TITLE
Update architecture overview

### PR DESCRIPTION
**Pull request by dongreenberg**

### CODE REVIEW

**Comment by dongreenberg** [comment]
* **#1036** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1036?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dongreenberg and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: docs/architecture.rst [hunk 1]
```diff
@@ -1,142 +1,211 @@
 Architecture Overview
 =====================
 
+Runhouse is a Python library that allows any application to flexibly and powerfully utilize remote compute
+infrastructure by deploying and calling remote services on the fly. It is principally designed for Machine
+Learning-style workloads (online, offline, training, and inference), where the need for heterogeneous
+remote compute is frequent and flexibility is paramount to minimize costs.
```
Comment (mkandler) [inline docs/architecture.rst:7]: Love this intro paragraph. When this is all finalized I think we may want to move some of this directly to the docs index page, as an [introduction](https://python.langchain.com/v0.2/docs/introduction/).

I don't think it (the first sections through "Why?") should be hidden in a secondary page in the "Other Topics" nav section

File: docs/architecture.rst [hunk 2]
```diff
@@ -1,142 +1,211 @@
 Architecture Overview
 =====================
 
+Runhouse is a Python library that allows any application to flexibly and powerfully utilize remote compute
+infrastructure by deploying and calling remote services on the fly. It is principally designed for Machine
```
Comment (py-rh) [inline docs/architecture.rst:5]: Nit - I'm not in love with "on the fly" which connotes ad hoc. 
"through a simple, programmable API"? Or something else is fine too.

File: docs/architecture.rst [hunk 3]
```diff
@@ -1,142 +1,211 @@
...
+Learning-style workloads (online, offline, training, and inference), where the need for heterogeneous
+remote compute is frequent and flexibility is paramount to minimize costs.
+
+The basic hypothesis is that incorporating dynamic compute into the runtime of an application, like workflow
```
Comment (py-rh) [inline docs/architecture.rst:9]: Could simplify? 
_Incorporating dynamic compute into workflow orchestrators  (Airflow, Prefect) or distributed libraries (Ray, Spark) is far more disruptive and less flexible at every level (dev workflow, debugging, DevOps, infra) than calling a remote service from those applications. _

File: docs/architecture.rst [hunk 4]
```diff
@@ -1,142 +1,211 @@
...
+orchestrators (Airflow, Prefect) or distributed libraries (Ray, Spark) is far more disruptive and less flexible at
+every level (dev workflow, debugging, DevOps, infra) than calling a remote service on the compute within the
+application. Calling a function or class as a remote service is a common pattern (e.g. microservices, Temporal),
+but divides the code into multiple applications which multiplies the DevOps overhead, each having their own
```
Comment (py-rh) [inline docs/architecture.rst:13]: nit, "but this"

File: docs/architecture.rst [hunk 5]
```diff
@@ -1,142 +1,211 @@
...
+flexibility in Python applications, without disrupting the runtime or cleaving the application, by offloading
+functions and classes to remote compute as services on the fly.
+
+Why?
+=====================
 
-
-Runhouse Resources
+This solves a few major problems for AI teams:
+1. **Cost**: The added flexibility Runhouse introduces to allocate compute only while needed, right-size instances based
```
Comment (py-rh) [inline docs/architecture.rst:22]: Sentence is a bit run on. 

Runhouse introduces the flexibility to allocate compute only while needed, right-size instances based on the workload, work across multiple regions or clouds for lower costs, and share compute and services across tasks. Users typically achieve significant cost savings, on the order of 50-75%, depending on the workload.

File: docs/architecture.rst [hunk 6]
```diff
@@ -1,142 +1,211 @@
...
+    on the workload, reach across regions or clouds for lower costs, and share compute and services across tasks
+    generally leads to large cost savings, on the order of 50-75% depending on the workload.
+2. **Development at scale**: Powerful hardware such as GPUs or distributed clusters (Spark, Ray) can be hugely
+    disruptive, requiring all development, debugging, automation, and deployment to occur on their runtime. Ray, Spark,
+    or PyTorch distributed users for example must be tunneled into the head node at all times for development, leading
+    to a proliferation of hosted notebook services as a stop-gap. Runhouse allows Python to orchestrate to these
+    systems remotely, returning the development workflow and operations to that of standard Python. Teams using Runhouse
```
Comment (py-rh) [inline docs/architecture.rst:29]: nit "to standard Python" 
omit "that of"

File: docs/architecture.rst [hunk 7]
```diff
@@ -1,142 +1,211 @@
...
 
-* **Configs**: Set and preserve default configs across environments.
+  .. note::
 
-Management
-----------
 
-Runhouse provides tools for visibility and management of resources as long-living assets shared by teams or projects.
+1. **Specify and/or Allocate Compute**: Runhouse can allocate compute to the application on the fly, either by
+    utilizing an existing VM or Ray cluster, or allocating one fresh using local cloud or K8s credentials. The
```
Comment (py-rh) [inline docs/architecture.rst:69]: nit: "allocating a new one using"

File: docs/architecture.rst [hunk 8]
```diff
@@ -1,142 +1,211 @@
...
+    `rh.cluster` constructor is generally used to specify and interact with remote compute, including bringing it up
+    if necessary (`cluster.up_if_not()`).
+2. **Starting the Runhouse Server Daemon**: If not already running, the client will start the Runhouse API server daemon
+    on the compute and form a network connection (either over SSH or HTTP/S). Dependencies can be specified to be
```
Comment (py-rh) [inline docs/architecture.rst:73]: nit "secure network connection" 
if you say it, people will believe it

File: docs/architecture.rst [hunk 9]
```diff
@@ -1,142 +1,211 @@
...
+        e. New workers can be constructed with `rh.env`, which specifies the details of the Python environment
+            (packages, environment variables) in which the process will be constructed. By default workers live
+            in the same Python environment as the daemon, but can also be started in a conda environment or a
+            separate node.
+3. **Deploying Functions or Classes**: The user specifies a function or class to be deployed to the remote compute
+    using the `rh.function` or `rh.module` constructors (or by subclassing `rh.Module`), and calling
+    `remote_obj = my_obj.to(my_cluster, worker=my_env)`. The Runhouse client library extracts the path, module name,
+    and importable name from the function or class. If the function or class is defined in local code, the repo or
+    package is rsynced onto the cluster. An instruction with the import path is sent to the cluster to
+    construct the function or class in a particular worker and assign it a name in the key-value store, perhaps
```
Comment (py-rh) [inline docs/architecture.rst:93]: "upserts it into the key-value store." might be more concise?

File: docs/architecture.rst (hunk 3 above)
Comment (mkandler) [reply docs/architecture.rst:9]: +1 here, I didn't have a solution but found this sentence too heavy and hard to parse

File: docs/architecture.rst (hunk 4 above)
Comment (mkandler) [inline docs/architecture.rst:13]: Maybe break up sentence with "multiple applications which multiples" -> "multiple applications. This increases DevOps overhead because each application has ..."

File: docs/architecture.rst [hunk 10]
```diff
@@ -1,142 +1,211 @@
...
+    gpu.down()
 
-* **Configs**: Set and preserve default configs across environments.
+  .. note::
 
-Management
-----------
 
-Runhouse provides tools for visibility and management of resources as long-living assets shared by teams or projects.
+1. **Specify and/or Allocate Compute**: Runhouse can allocate compute to the application on the fly, either by
```
Comment (mkandler) [inline docs/architecture.rst:68]: UX suggestion: Use H3's for these instead of bullets so they show up in the Table of Contents for the page. Will also make the sublist in "Starting the daemon" just a normal list.

File: docs/architecture.rst [hunk 11]
```diff
@@ -1,142 +1,211 @@
...
+            objects in memory, and exposing an HTTP API to call methods on those objects by name.
+        b. The objects are held in a single default worker process by default, but can be sent to other worker
+            processes, including on other nodes in the cluster, to achieve powerful parallelism out of the box.
+        c. If I call GET http://myserver:32300/my_object/my_method, the daemon will look up the object named
+            "my_object", issue an instruction for its worker to call the method "my_method" on it, and
+            return the result.
+        d. The HTTP server and workers can handle thousands of concurrent calls per second, and have similar latency
+            under simple conditions to Flask.
+        e. New workers can be constructed with `rh.env`, which specifies the details of the Python environment
+            (packages, environment variables) in which the process will be constructed. By default workers live
```
Comment (mkandler) [inline docs/architecture.rst:85]: nit: "By default, workers" (otherwise I read as "default workers" whatever those are)

File: docs/architecture.rst [hunk 12]
```diff
@@ -1,142 +1,211 @@
...
+runtime (which is just Python) or orchestrating *between* clusters (even of different types).
+For this reason, it has no other runtime to setup than Python itself, can be used to orchestrate your distributed code so you
+can use your Ray or Spark clusters less disruptively within your stack (e.g. sending a function which uses
+Ray over to the head node of the Ray cluster, where the Ray will execute as usual).
+
+<diagram showing Ray calling actors within the cluster, Runhouse calling outside into other processes, VMs,
+Ray clusters, K8s clusters, clouds>
+
+This also fixes some other sharp
+edges with such systems which significantly reduces costs, such as utilizing more than one cluster in an application,
```
Comment (py-rh) [inline docs/architecture.rst:136]: This also fixes sharp edges and inefficiencies with these systems, such as the inability to use more than one cluster in an application or sharing a cluster between multiple callers.

File: docs/architecture.rst [hunk 13]
```diff
@@ -1,142 +1,211 @@
...
+~~~~~~~~~~~~~~~~~~
 
-* `Runhouse Den <https://www.run.house/dashboard>`_: provides an individual or admin view of all resources, secrets,
-  groups, and sharing. Resource metadata is automatically versioned in RNS, allowing teams to maintain single-sources
-  of truth for assets with zero downtime to update or roll back, and trace exact lineage for any existing resource.
+Runhouse's APIs bear similarity to other systems, so it's helpful to compare and contrast. In many cases,
+Runhouse is not a replacement for these systems, but rather a complement or extension. In others, you may be able
+to replace your usage of the other system entirely with Runhouse.
+
+*Distributed frameworks* like Ray, Spark, or Elixr make it possible to offload execution onto separate
```
Comment (mkandler) [inline docs/architecture.rst:124]: These could also potentially be h3s to break up the section and improve readability

File: docs/architecture.rst [hunk 14]
```diff
@@ -1,142 +1,211 @@
...
+
+Generally, workflow orchestrators are built to be good at monitoring, telemetry, fault-tolerance, and scheduling, so
+we recommend using one strictly for those features and using Runhouse within your pipeline nodes for the heterogeneous
+compute and remote execution. You can also save a lot of money by reusing compute across multiple nodes or reusing
+services across multiple pipelines with Runhouse, which is generally not possible with workflow orchestrators.
+
+<diagram showing an Airflow pipeline calling multiple steps on one cluster, and another pipeline calling a service
+created by the first>
+
+*Serverless frameworks* like Modal or AWS Lambda allow for the allocation of services on the fly, but within a well
```
Comment (py-rh) [inline docs/architecture.rst:158]: nit well-defined

File: docs/architecture.rst [hunk 15]
```diff
@@ -1,142 +1,211 @@
...
+to offload execution within the application (though you could call a predefined script entrypoint or API
+endpoint). Runhouse uses SkyPilot to allocate compute, but is vertically integrated to be able
+to perform allocation, (re)deployment, and management of a new service all in Python so the new compute can be used
+instantly within the existing application. It also doesn't need to perform allocation to create new services -
+it can use existing compute or static VMs.
+
+<diagram showing SkyPilot running a CLI command on the remote compute and getting back logs, and Runhouse using
+SkyPilot to launch new compute, send a function to it, and call it>
+
+*GPU/Accelerator dispatch* systems like PyTorch, Jax, or Mojo give the ability to offload computation to a local GPU or
```
Comment (py-rh) [inline docs/architecture.rst:176]: Do we need this? Is it unclear the difference btw Runhouse and PyTorch?

File: docs/architecture.rst [hunk 16]
```diff
@@ -1,142 +1,211 @@
...
+
+Saving, Loading, and Sharing
+~~~~~~~~~~~~~~~~~~
 
-* **Organization structure**: Both resources and users can be organized into organizations to apply and manage shared
-  access and permissions. Resources can be created for an organization by specifying its name in full
-  when creating or loading down a Runhouse resource.
+Runhouse resources (clusters, functions, modules, environments) can be saved, shared, and reused based on a compact
+JSON metadata signature. This allows for easy sharing of clusters and services across users and environments, which
+can often lead to massive cost savings. The `Den <https://www.run.house/dashboard>`_ hosted metadata store can be
```
Comment (mkandler) [inline docs/architecture.rst:188]: I think we could expand on what Den is here. A full sentence explaining before saying how to access maybe?

File: docs/architecture.rst [hunk 17]
```diff
@@ -1,142 +1,211 @@
...
+
+*Serverless frameworks* like Modal or AWS Lambda allow for the allocation of services on the fly, but within a well
+defined sandbox, and not strictly from within regular Python - they require specific pre-packaging or CLI launch
+commands outside Python. Runhouse runs fully in a Python interpreter so it can extend the compute power of practically
+any existing Python application, and allocates services inside your own compute, wherever that may be. We may even
+support serverless systems as compute backends in the future.
+
+<diagram showing Modal creating new services within their own compute, and Runhouse within EC2, K8s, GCP, etc.>
+
+*Infrastructure in code* tools like SkyPilot or Pulumi allocate compute on the fly, but can't utilize it instantly
```
Comment (mkandler) [inline docs/architecture.rst:166]: A very general nit: Several sentences in this doc follow this format and don't need the comma before "but" (unless the second half is a standalone sentence or a the comma comes from an inserted thought)

File: docs/architecture.rst [hunk 18]
```diff
@@ -1,142 +1,211 @@
...
+            versioning mismatch errors between local and remote package versions.
+        c. From here on, you can think of Runhouse as facilitating
+            regular object-oriented programming but with the objects living remotely, maybe in a different cluster,
+            region, or cloud than the local code.
+        d. Python behavior like async, exceptions, printing, and logging are all preserved across remote calls, but
+            can be disabled or controlled if desired.
+5. **Saving and Loading**: The Runhouse client can save and load objects to and from the local filesystem, or to a
+    remote metadata store. This allows for easy sharing of clusters and services across users and environments,
+    and for versioning and rollback of objects. The metadata store can be accessed from any Python interpreter,
+    and is backed by a management API to view and manage all resources.
```
Comment (jlewitt1) [inline docs/architecture.rst:111]: i know we had this before, but "management API" makes it sound more like REST API reference docs, rather than something like a "monitoring or management portal"

File: docs/architecture.rst [hunk 19]
```diff
@@ -1,142 +1,211 @@
...
+6. **Terminating Modules, Workers, or Clusters**: When a remote object is no longer needed, it can be deallocated from
```
Comment (jlewitt1) [inline docs/architecture.rst:112]: thoughts on adding a note about autostop here? (or linking out to where we go into detail about it)

File: docs/architecture.rst [hunk 20]
```diff
@@ -1,142 +1,211 @@
...
+  .. code-block:: python
 
-Accessibility and Management
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    import runhouse as rh
 
-Runhouse enables the portability and sharing of resources across users and environments, and provides
-tools for visibility and management of these resources as long-living assets shared by teams or projects.
+    # [1] and [2]
+    gpu = rh.cluster(name="rh-a10x", instance_type="A10G:1", provider="aws").up_if_not()
```
Comment (jlewitt1) [inline docs/architecture.rst:47]: wdyt about using "cheapest" here to highlight multi cloud / cost savings

File: docs/architecture.rst (hunk 9 above)
Comment (jlewitt1) [reply docs/architecture.rst:93]: + 1 feels more like an implementation detail - at a higher level this also shows how easy it is to update and redeploy an existing service

File: docs/architecture.rst [hunk 21]
```diff
@@ -1,142 +1,243 @@
...
+flexibility in Python without disrupting the runtime or cleaving the application, by offloading
+functions and classes to remote compute as services on the fly.
+
+Why?
+----
+
+This solves a few major problems for AI teams:
+
+#. **Cost**: Runhouse introduces the flexibility to allocate compute only while needed, right-size instances based on
+   the size of the workload, work across multiple regions or clouds for lower costs, and share compute and services
```
Comment (carolineechen) [inline docs/architecture.rst:26]: doesn't read super smooth - 

```suggestion
#. **Cost**: Runhouse introduces the flexibility to allocate only the compute you need when you need it.
   You can work across multiple regions or clouds for lower costs, and share compute and services
```

File: docs/architecture.rst [hunk 22]
```diff
@@ -1,142 +1,243 @@
...
+   under simple conditions to Flask.
+#. New workers can be constructed with ``rh.env``, which specifies the details of the Python environment
+   (packages, environment variables) in which the process will be constructed. By default, workers live
+   in the same Python environment as the daemon but can also be started in a conda environment or a
+   separate node.
+
+3. Deploying Functions or Classes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The user specifies a function or class to be deployed to the remote compute
+using the ``rh.function``` or ``rh.module`` constructors (or by subclassing ``rh.Module``), and calling
```
Comment (carolineechen) [inline docs/architecture.rst:101]: extra tick here

```suggestion
using the ``rh.function`` or ``rh.module`` constructors (or by subclassing ``rh.Module``), and calling
```

File: docs/architecture.rst [hunk 23]
```diff
@@ -1,142 +1,243 @@
...
+   "my_object", issue an instruction for its worker to call the method "my_method" on it, and
+   return the result.
+#. The HTTP server and workers can handle thousands of concurrent calls per second, and have similar latency
+   under simple conditions to Flask.
+#. New workers can be constructed with ``rh.env``, which specifies the details of the Python environment
+   (packages, environment variables) in which the process will be constructed. By default, workers live
+   in the same Python environment as the daemon but can also be started in a conda environment or a
+   separate node.
+
+3. Deploying Functions or Classes
```
Comment (carolineechen) [inline docs/architecture.rst:98]: in the snippet above we included `rh.env` in [3] and not [2]. should we briefly explain the env setup here as well?

File: docs/architecture.rst [hunk 24]
```diff
@@ -1,142 +1,243 @@
...
+    sd_upsampler = rh.function(name="/my_username/sd_upsampler")
+    high_res_imgs = sd_upsampler(imgs)
+
+    # [6]
+    gpu.teardown()
+
+.. note::
+
+
+1. Specify and/or Allocate Compute
```
Comment (carolineechen) [inline docs/architecture.rst:71]: the length of these explanations is long enough that one would need to scroll back and forth between the code snippets. thoughts on rewriting the code snippet corresponding to the explanation below each number too?



**#. header**
```
gpu = rh.cluster(name="rh-a10x", instance_type="A10G:1", provider="aws").up_if_not()
```
< desc >

File: docs/architecture.rst [hunk 25]
```diff
@@ -1,142 +1,243 @@
...
+GPU/Accelerator dispatch (e.g. PyTorch, Jax, Mojo)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+GPU/Accelerator dispatch systems give the ability to offload computation to a local GPU or
+TPU. Runhouse does not have this capability but can offload a function or class to a remote instance with an
+accelerator, which can then itself use libraries like PyTorch or Jax (and maybe one day Mojo) to use the accelerator.
+
+<diagram showing PyTorch dispatching to the GPU within the node, Runhouse dispatching to a remote node which then uses
+PyTorch to dispatch to the GPU>
+
+Saving, Loading, and Sharing
```
Comment (carolineechen) [inline docs/architecture.rst:214]: some overlaps with # 5
